File: pocketr-java.api/src/main/java/com/decrux/pocketr/api/config/security/SecurityConfig.java
package com.decrux.pocketr.api.config.security;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.csrf.CookieCsrfTokenRepository;

@Configuration
@EnableWebSecurity
public class SecurityConfig {

    private final SpaCsrfTokenRequestHandler spaCsrfTokenRequestHandler;
    private final CustomAuthenticationSuccessHandler customAuthenticationSuccessHandler;
    private final CustomAuthenticationFailureHandler customAuthenticationFailureHandler;
    private final CustomLogoutSuccessHandler customLogoutSuccessHandler;
    private final CustomAuthenticationEntryPoint customAuthenticationEntryPoint;
    private final CustomUserDetailsService customUserDetailsService;
    private final String csrfCookiePath;

    public SecurityConfig(
            SpaCsrfTokenRequestHandler spaCsrfTokenRequestHandler,
            CustomAuthenticationSuccessHandler customAuthenticationSuccessHandler,
            CustomAuthenticationFailureHandler customAuthenticationFailureHandler,
            CustomLogoutSuccessHandler customLogoutSuccessHandler,
            CustomAuthenticationEntryPoint customAuthenticationEntryPoint,
            CustomUserDetailsService customUserDetailsService,
            @Value("${app.security.csrf-cookie-path:}") String csrfCookiePath
    ) {
        this.spaCsrfTokenRequestHandler = spaCsrfTokenRequestHandler;
        this.customAuthenticationSuccessHandler = customAuthenticationSuccessHandler;
        this.customAuthenticationFailureHandler = customAuthenticationFailureHandler;
        this.customLogoutSuccessHandler = customLogoutSuccessHandler;
        this.customAuthenticationEntryPoint = customAuthenticationEntryPoint;
        this.customUserDetailsService = customUserDetailsService;
        this.csrfCookiePath = csrfCookiePath;
    }

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
        CookieCsrfTokenRepository csrfTokenRepository = CookieCsrfTokenRepository.withHttpOnlyFalse();
        if (!csrfCookiePath.isBlank()) {
            csrfTokenRepository.setCookiePath(csrfCookiePath);
        }

        return http
                .csrf(csrf -> csrf
                        .csrfTokenRepository(csrfTokenRepository)
                        .csrfTokenRequestHandler(spaCsrfTokenRequestHandler)
                )
                .sessionManagement(session -> session.maximumSessions(1))
                .formLogin(form -> form
                        .loginPage(SecurityConstants.FRONTEND_LOGIN_URL.getValue())
                        .permitAll()
                        .loginProcessingUrl(SecurityConstants.BACKEND_LOGIN_PROCESSING_URL.getValue())
                        .permitAll()
                        .usernameParameter("email")
                        .successHandler(customAuthenticationSuccessHandler)
                        .failureHandler(customAuthenticationFailureHandler)
                )
                .authorizeHttpRequests(authorizeRequests -> authorizeRequests
                        .requestMatchers(SecurityConstants.FRONTEND_STATIC_ASSETS.getValue()).permitAll()
                        .requestMatchers(SecurityConstants.FRONTEND_FAVICON.getValue()).permitAll()
                        .requestMatchers(SecurityConstants.FRONTEND_LANDING_PAGE.getValue()).permitAll()
                        .requestMatchers(SecurityConstants.FRONTEND_LOGIN_URL.getValue()).permitAll()
                        .requestMatchers(SecurityConstants.FRONTEND_REGISTER_URL.getValue()).permitAll()
                        .requestMatchers(SecurityConstants.BACKEND_REGISTER_ENDPOINT.getValue()).permitAll()
                        .requestMatchers(SecurityConstants.GET_CSRF_TOKEN_ENDPOINT.getValue()).permitAll()
                        .anyRequest().authenticated()
                )
                .exceptionHandling(exception -> exception
                        .authenticationEntryPoint(customAuthenticationEntryPoint)
                )
                .userDetailsService(customUserDetailsService)
                .logout(logout -> logout
                        .logoutUrl(SecurityConstants.BACKEND_LOGOUT_PROCESSING_URL.getValue())
                        .clearAuthentication(true)
                        .invalidateHttpSession(true)
                        .logoutSuccessHandler(customLogoutSuccessHandler)
                )
                .build();
    }

    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }
}

File: pocketr-api/src/main/kotlin/com/decrux/pocketr_api/config/security/SecurityConfig.kt
package com.decrux.pocketr_api.config.security

import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.csrf.CookieCsrfTokenRepository

@Configuration
@EnableWebSecurity
class SecurityConfig(
    private val spaCsrfTokenRequestHandler: SpaCsrfTokenRequestHandler,
    private val customAuthenticationSuccessHandler: CustomAuthenticationSuccessHandler,
    private val customAuthenticationFailureHandler: CustomAuthenticationFailureHandler,
    private val customLogoutSuccessHandler: CustomLogoutSuccessHandler,
    private val customAuthenticationEntryPoint: CustomAuthenticationEntryPoint,
    private val customUserDetailsService: CustomUserDetailsService,
    @Value("\${app.security.csrf-cookie-path:}")
    private val csrfCookiePath: String,
) {
    @Bean
    fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
        val csrfTokenRepository =
            CookieCsrfTokenRepository.withHttpOnlyFalse().apply {
                if (csrfCookiePath.isNotBlank()) {
                    setCookiePath(csrfCookiePath)
                }
            }

        return http
            .csrf { csrf ->
                csrf
                    .csrfTokenRepository(csrfTokenRepository)
                    .csrfTokenRequestHandler(spaCsrfTokenRequestHandler)
            }.sessionManagement { session ->
                session.maximumSessions(1)
            }.formLogin { form ->
                form
                    .loginPage(SecurityConstants.FRONTEND_LOGIN_URL.value)
                    .permitAll()
                    .loginProcessingUrl(SecurityConstants.BACKEND_LOGIN_PROCESSING_URL.value)
                    .permitAll()
                    .usernameParameter("email")
                    .successHandler(customAuthenticationSuccessHandler)
                    .failureHandler(customAuthenticationFailureHandler)
            }.authorizeHttpRequests { authorizeRequests ->
                authorizeRequests
                    .requestMatchers(SecurityConstants.FRONTEND_STATIC_ASSETS.value)
                    .permitAll()
                    .requestMatchers(SecurityConstants.FRONTEND_FAVICON.value)
                    .permitAll()
                    .requestMatchers(SecurityConstants.FRONTEND_LANDING_PAGE.value)
                    .permitAll()
                    .requestMatchers(SecurityConstants.FRONTEND_LOGIN_URL.value)
                    .permitAll()
                    .requestMatchers(SecurityConstants.FRONTEND_REGISTER_URL.value)
                    .permitAll()
                    .requestMatchers(SecurityConstants.BACKEND_REGISTER_ENDPOINT.value)
                    .permitAll()
                    .requestMatchers(SecurityConstants.GET_CSRF_TOKEN_ENDPOINT.value)
                    .permitAll()
                    .anyRequest()
                    .authenticated()
            }.exceptionHandling { exception ->
                exception.authenticationEntryPoint(customAuthenticationEntryPoint)
            }.userDetailsService(customUserDetailsService)
            .logout { logout ->
                logout
                    .logoutUrl(SecurityConstants.BACKEND_LOGOUT_PROCESSING_URL.value)
                    .clearAuthentication(true)
                    .invalidateHttpSession(true)
                    .logoutSuccessHandler(customLogoutSuccessHandler)
            }.build()
    }

    @Bean
    fun passwordEncoder(): PasswordEncoder = BCryptPasswordEncoder()
}
